File: frontend/src/types/bot.ts
/**
 * Bot Platform Types
 */

// ============================================================================
// Bot Core Types
// ============================================================================

export interface Bot {
  id: string;
  name: string;
  username: string;  // @bot_username format
  description?: string;
  avatar_url?: string;
  owner_id: string;
  is_active: boolean;
  handler_type: BotHandlerType;
  webhook_url?: string;
  command_count: number;
  created_at: string;
  /** Bot token (only shown once after creation) */
  token?: string;
}

export enum BotHandlerType {
  Internal = 'internal',
  Webhook = 'webhook',
  AI = 'ai',
}

export interface BotCommand {
  id: string;
  bot_id: string;
  command: string;       // e.g. "/start"
  description?: string;
  handler_type: BotHandlerType;
  handler_url?: string;
  response_template?: string;  // For simple internal bots
}

export interface BotWebhook {
  id: string;
  bot_id: string;
  url: string;
  events: BotEventType[];
  secret: string;
  active: boolean;
  last_triggered_at?: string;
  last_status?: number;
}

export type BotEventType =
  | 'message'
  | 'command'
  | 'join'
  | 'leave'
  | 'edit'
  | 'delete';

// ============================================================================
// Bot Session / FSM
// ============================================================================

export interface BotSession {
  bot_id: string;
  user_id: string;
  chat_id: string;
  state: string;          // FSM state identifier
  context: Record<string, unknown>;  // Arbitrary context data
  updated_at: string;
}

// ============================================================================
// Bot Event (what gets dispatched to bots)
// ============================================================================

export interface BotEvent {
  event_type: BotEventType;
  bot_id: string;
  user_id: string;
  chat_id: string;
  message?: {
    id: string;
    content: string;
    is_command: boolean;
    command?: string;     // e.g. "/start"
    args?: string;        // e.g. "arg1 arg2"
  };
  timestamp: string;
}

// ============================================================================
// Bot API Response (for external webhooks)
// ============================================================================

export interface BotApiResponse {
  /** Message to send back */
  text?: string;
  /** Sticker ID to send */
  sticker_id?: string;
  /** Keyboard buttons (inline) */
  inline_keyboard?: InlineKeyboardButton[][];
  /** Delete the triggering message */
  delete_message?: boolean;
  /** Edit the triggering message */
  edit_message?: string;
}

export interface InlineKeyboardButton {
  text: string;
  callback_data?: string;
  url?: string;
  switch_inline_query?: string;
}

// ============================================================================
// Bot Store (Marketplace)
// ============================================================================

export interface BotStoreListing {
  id: string;
  name: string;
  username: string;
  description: string;
  avatar_url?: string;
  category: BotCategory;
  rating: number;
  install_count: number;
  is_verified: boolean;
  is_premium: boolean;
  author: string;
  commands: string[];  // Sample commands
}

export enum BotCategory {
  Utility = 'utility',
  Entertainment = 'entertainment',
  AI = 'ai',
  Moderation = 'moderation',
  Integration = 'integration',
  Games = 'games',
}

// ============================================================================
// CRUD Payloads
// ============================================================================

export interface CreateBotPayload {
  name: string;
  username: string;
  description?: string;
  handler_type: BotHandlerType;
  webhook_url?: string;
  ai_prompt?: string;  // For AI bots
}

export interface UpdateBotPayload {
  name?: string;
  description?: string;
  avatar_url?: string;
  webhook_url?: string;
  is_active?: boolean;
}

export interface CreateCommandPayload {
  command: string;
  description?: string;
  handler_type: BotHandlerType;
  handler_url?: string;
  response_template?: string;
}

export interface CreateWebhookPayload {
  url: string;
  events: BotEventType[];
  secret?: string;
}

File: server/src/models/bot.rs
//! Bot Platform Models

use serde::{Deserialize, Serialize};
use sqlx::FromRow;

// ============================================================================
// Database Models
// ============================================================================

#[derive(Debug, Clone, FromRow, Serialize)]
pub struct Bot {
    pub id: String,
    pub name: String,
    pub username: String,
    pub description: Option<String>,
    pub avatar_url: Option<String>,
    pub owner_id: String,
    pub handler_type: String,
    pub webhook_url: Option<String>,
    pub ai_prompt: Option<String>,
    pub is_active: bool,
    #[serde(default)]
    pub command_count: i64,
    pub created_at: String,
}

#[derive(Debug, Clone, FromRow, Serialize)]
pub struct BotCommand {
    pub id: String,
    pub bot_id: String,
    pub command: String,
    pub description: Option<String>,
    pub handler_type: String,
    pub handler_url: Option<String>,
    pub response_template: Option<String>,
    pub created_at: String,
}

#[derive(Debug, Clone, FromRow, Serialize)]
pub struct BotWebhook {
    pub id: String,
    pub bot_id: String,
    pub url: String,
    pub events: String, // JSON array
    pub secret: String,
    pub active: bool,
    pub last_triggered_at: Option<String>,
    pub last_status: Option<i32>,
    pub created_at: String,
}

#[derive(Debug, Clone, FromRow, Serialize, Deserialize)]
#[allow(dead_code)]
pub struct BotSession {
    pub bot_id: String,
    pub user_id: String,
    pub chat_id: String,
    pub state: String,
    pub context: String, // JSON
    pub updated_at: String,
}

#[derive(Debug, Clone, FromRow, Serialize)]
pub struct BotStoreListing {
    pub id: String,
    pub name: String,
    pub username: String,
    pub description: String,
    pub avatar_url: Option<String>,
    pub category: String,
    pub rating: f64,
    pub install_count: i64,
    pub is_verified: bool,
    pub is_premium: bool,
    pub author: String,
    pub commands: String, // JSON array
    pub bot_id: Option<String>,
}

// ============================================================================
// Request/Response Types
// ============================================================================

#[derive(Debug, Deserialize)]
pub struct CreateBotRequest {
    pub name: String,
    pub username: String,
    pub description: Option<String>,
    pub handler_type: Option<String>,
    pub webhook_url: Option<String>,
    pub ai_prompt: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct UpdateBotRequest {
    pub name: Option<String>,
    pub description: Option<String>,
    pub avatar_url: Option<String>,
    pub webhook_url: Option<String>,
    pub is_active: Option<bool>,
}

#[derive(Debug, Deserialize)]
pub struct CreateCommandRequest {
    pub command: String,
    pub description: Option<String>,
    pub handler_type: Option<String>,
    pub handler_url: Option<String>,
    pub response_template: Option<String>,
}

#[derive(Debug, Deserialize)]
pub struct CreateWebhookRequest {
    pub url: String,
    pub events: Option<Vec<String>>,
    pub secret: Option<String>,
}

#[derive(Debug, Serialize)]
pub struct BotTokenResponse {
    pub token: String,
}

#[derive(Debug, Serialize)]
pub struct BotWithToken {
    #[serde(flatten)]
    pub bot: Bot,
    pub token: String, // Plain text token (only sent once)
}

// ============================================================================
// Bot Event (dispatched to bots)
// ============================================================================

#[derive(Debug, Clone, Serialize, Deserialize)]
#[allow(dead_code)]
pub struct BotEvent {
    pub event_type: String,
    pub bot_id: String,
    pub user_id: String,
    pub chat_id: String,
    pub message: Option<EventMessage>,
    pub timestamp: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[allow(dead_code)]
pub struct EventMessage {
    pub id: String,
    pub content: String,
    pub is_command: bool,
    pub command: Option<String>,
    pub args: Option<String>,
}

// ============================================================================
// Bot API Response (what external webhooks return)
// ============================================================================

#[derive(Debug, Serialize, Deserialize)]
#[allow(dead_code)]
pub struct BotApiResponse {
    pub text: Option<String>,
    pub sticker_id: Option<String>,
    pub inline_keyboard: Option<Vec<Vec<InlineKeyboardButton>>>,
    pub delete_message: Option<bool>,
    pub edit_message: Option<String>,
}

#[derive(Debug, Serialize, Deserialize)]
#[allow(dead_code)]
pub struct InlineKeyboardButton {
    pub text: String,
    pub callback_data: Option<String>,
    pub url: Option<String>,
    pub switch_inline_query: Option<String>,
}
